fdbserver/IPager.h

```diff
@@ -97,67 +151,341 @@ class ArenaPage : public ReferenceCounted<ArenaPage>, public FastAllocated<Arena
 		}
 	}
 
-	uint8_t const* begin() const { return (uint8_t*)buffer; }
+	// Before using these, either init() or postRead1 and postRead2() must be called
+	const uint8_t* data() const { return pPayload; }
+	uint8_t* mutateData() const { return (uint8_t*)pPayload; }
+	int dataSize() const { return payloadSize; }
+
+	const uint8_t* rawData() const { return buffer; }
+	uint8_t* rawData() { return buffer; }
+	int rawSize() const { return bufferSize; }
+
+#pragma pack(push, 1)
+
+	// The next few structs describe the byte-packed physical structure.  The fields of Page
+	// cannot change, but new header versions and encoding types can be added and existing
+	// header versions and encoding type headers could change size as offset information
+	// is stored to enable efficient jumping to the encoding header or payload.
+	struct Page {
+		uint8_t headerVersion;
+		EncodingType encodingType;
+
+		// Encoding header comes after main header
+		uint8_t encodingHeaderOffset;
+
+		// Payload comes after encoding header
+		uint8_t payloadOffset;
+
+		// Get main header pointer, casting to its type
+		template <typename T>
+		T* getMainHeader() const {
+			return (T*)(this + 1);
+		}
 
-	uint8_t* mutate() { return (uint8_t*)buffer; }
+		// Get encoding header pointer, casting to its type
+		template <typename T>
+		T* getEncodingHeader() const {
+			return (T*)((uint8_t*)this + encodingHeaderOffset);
+		}
+
+		// Get payload pointer
+		uint8_t* getPayload() const { return (uint8_t*)this + payloadOffset; }
+	};
+
+	// Header version 1
+	// The header is responsible for protecting all Page bytes outside of the payload.
+	// HeaderV1 does so with a 64-bit XXHash checksum
+	//
+	// Most other fields are forensic in nature and are not required to be set for correct
+	// behavior but they can faciliate forensic investigation of data on disk.  Some of them
+	// could be used for sanity checks at runtime.
+	struct HeaderV1 {
+		PageType pageType;
+		// The meaning of pageSubType is based on pageType
+		//   For Queue pages, pageSubType is QueueID
+		//   For BTree nodes, pageSubType is Height (also stored in BTreeNode)
+		uint8_t pageSubType;
+		XXH64_hash_t checksum;
+
+		// Physical page ID of first block on disk of the ArenaPage
+		PhysicalPageID firstPhysicalPageID;
+		// The first logical page ID the ArenaPage was referenced by when last written
+		LogicalPageID lastKnownLogicalPageID;
+		// The first logical page ID of the parent of this ArenaPage when last written
+		LogicalPageID lastKnownParentLogicalPageID;
+
+		// Time and write version as of the last update to this page.
+		// Note that for relocated pages, writeVersion should not be updated.
+		double writeTime;
+		Version writeVersion;
+
+		// Update checksum
+		void update(uint8_t* headerBytes, int len) {
```

Comment:
nit: "updateChecksum" and "verifyChecksum" as more explicit names?

Reply:
Will change.